File: rust/benches/throughput.rs
//! Sustained two-thread throughput, this crate vs rtrb vs crossbeam's ArrayQueue.
//!
//! Both threads are pinned to distinct physical cores on the same CCD, and the
//! channel allocation and thread spawn happen before the timed region (a barrier
//! starts the clock only once both sides are ready). Reports the median of 15 runs.

use std::sync::{Arc, Barrier};
use std::thread;
use std::time::Instant;

const ITEMS: u64 = 50_000_000;
const CAPACITY: usize = 1024;
const BATCH: usize = 64;
const PRODUCER_CORE: usize = 0;
const CONSUMER_CORE: usize = 2;
const RUNS: usize = 15;

fn pin(core: usize) {
    if let Some(ids) = core_affinity::get_core_ids() {
        if let Some(id) = ids.into_iter().find(|c| c.id == core) {
            core_affinity::set_for_current(id);
        }
    }
}

fn median_mops(mut samples: Vec<f64>) -> f64 {
    samples.sort_by(|a, b| a.partial_cmp(b).unwrap());
    samples[samples.len() / 2]
}

fn run_spsc_ring() -> f64 {
    let (mut tx, mut rx) = spsc_ring::channel::<u64>(CAPACITY);
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            for i in 0..ITEMS {
                while tx.push(i).is_err() {
                    std::hint::spin_loop();
                }
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    while received < ITEMS {
        if rx.pop().is_some() {
            received += 1;
        } else {
            std::hint::spin_loop();
        }
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

fn run_spsc_ring_batch() -> f64 {
    let (mut tx, mut rx) = spsc_ring::channel::<u64>(CAPACITY);
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            let mut chunk = [0u64; BATCH];
            let mut next = 0u64;
            while next < ITEMS {
                let want = ((ITEMS - next) as usize).min(BATCH);
                for (i, c) in chunk[..want].iter_mut().enumerate() {
                    *c = next + i as u64;
                }
                let mut sent = 0;
                while sent < want {
                    let n = tx.push_slice(&chunk[sent..want]);
                    if n == 0 {
                        std::hint::spin_loop();
                    }
                    sent += n;
                }
                next += want as u64;
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    let mut buf = [0u64; BATCH];
    while received < ITEMS {
        let n = rx.pop_slice(&mut buf);
        if n == 0 {
            std::hint::spin_loop();
        }
        received += n as u64;
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

// Same data movement as the slice API: values staged in a local buffer first.
fn run_rtrb_chunk_staged() -> f64 {
    let (mut tx, mut rx) = rtrb::RingBuffer::<u64>::new(CAPACITY);
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            let mut next = 0u64;
            while next < ITEMS {
                let want = ((ITEMS - next) as usize).min(BATCH).min(tx.slots());
                if want == 0 {
                    std::hint::spin_loop();
                    continue;
                }
                let mut staged = [0u64; BATCH];
                for (i, s) in staged[..want].iter_mut().enumerate() {
                    *s = next + i as u64;
                }
                let chunk = tx.write_chunk_uninit(want).unwrap();
                next += chunk.fill_from_iter(staged[..want].iter().copied()) as u64;
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    let mut buf = [0u64; BATCH];
    while received < ITEMS {
        let want = rx.slots().min(BATCH);
        if want == 0 {
            std::hint::spin_loop();
            continue;
        }
        let chunk = rx.read_chunk(want).unwrap();
        let (a, b) = chunk.as_slices();
        buf[..a.len()].copy_from_slice(a);
        buf[a.len()..a.len() + b.len()].copy_from_slice(b);
        received += (a.len() + b.len()) as u64;
        chunk.commit_all();
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

// rtrb's idiomatic chunk use: values are constructed directly in the ring, no
// staging pass. A slice API cannot express this; see the README tradeoff note.
fn run_rtrb_chunk_inplace() -> f64 {
    let (mut tx, mut rx) = rtrb::RingBuffer::<u64>::new(CAPACITY);
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            let mut next = 0u64;
            while next < ITEMS {
                let want = ((ITEMS - next) as usize).min(BATCH).min(tx.slots());
                if want == 0 {
                    std::hint::spin_loop();
                    continue;
                }
                let chunk = tx.write_chunk_uninit(want).unwrap();
                next += chunk.fill_from_iter(next..) as u64;
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    let mut buf = [0u64; BATCH];
    while received < ITEMS {
        let want = rx.slots().min(BATCH);
        if want == 0 {
            std::hint::spin_loop();
            continue;
        }
        let chunk = rx.read_chunk(want).unwrap();
        let (a, b) = chunk.as_slices();
        buf[..a.len()].copy_from_slice(a);
        buf[a.len()..a.len() + b.len()].copy_from_slice(b);
        received += (a.len() + b.len()) as u64;
        chunk.commit_all();
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

fn run_rtrb() -> f64 {
    let (mut tx, mut rx) = rtrb::RingBuffer::<u64>::new(CAPACITY);
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            for i in 0..ITEMS {
                while tx.push(i).is_err() {
                    std::hint::spin_loop();
                }
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    while received < ITEMS {
        if rx.pop().is_ok() {
            received += 1;
        } else {
            std::hint::spin_loop();
        }
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

fn run_crossbeam() -> f64 {
    let queue = Arc::new(crossbeam_queue::ArrayQueue::<u64>::new(CAPACITY));
    let gate = Arc::new(Barrier::new(2));
    let producer = {
        let queue = Arc::clone(&queue);
        let gate = Arc::clone(&gate);
        thread::spawn(move || {
            pin(PRODUCER_CORE);
            gate.wait();
            for i in 0..ITEMS {
                while queue.push(i).is_err() {
                    std::hint::spin_loop();
                }
            }
        })
    };
    pin(CONSUMER_CORE);
    gate.wait();
    let start = Instant::now();
    let mut received = 0u64;
    while received < ITEMS {
        if queue.pop().is_some() {
            received += 1;
        } else {
            std::hint::spin_loop();
        }
    }
    let secs = start.elapsed().as_secs_f64();
    producer.join().unwrap();
    ITEMS as f64 / secs / 1e6
}

fn report(name: &str, run: fn() -> f64) {
    let samples = (0..RUNS).map(|_| run()).collect::<Vec<_>>();
    println!("  {name:<22} {:8.1} Melem/s", median_mops(samples));
}

fn main() {
    println!(
        "throughput (median of {RUNS}, {ITEMS} items, capacity {CAPACITY}, cores {PRODUCER_CORE}->{CONSUMER_CORE})"
    );
    report("spsc-ring", run_spsc_ring);
    report("spsc-ring batch(64)", run_spsc_ring_batch);
    report("rtrb", run_rtrb);
    report("rtrb chunk(64) staged", run_rtrb_chunk_staged);
    report("rtrb chunk(64) in-place", run_rtrb_chunk_inplace);
    report("crossbeam-ArrayQueue", run_crossbeam);
}

File: cpp/bench/throughput.cpp
// Sustained two-thread throughput, this library vs rigtorp::SPSCQueue.
//
// Both threads are pinned to distinct physical cores on the same CCD; the queue
// allocation and thread spawn happen before the timed region (a barrier starts
// the clock once both sides are ready). Reports the median of 15 runs.

#include <algorithm>
#include <immintrin.h>
#include <barrier>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <thread>
#include <vector>

#include <rigtorp/SPSCQueue.h>

#include "bench_util.hpp"
#include "spsc/ring_buffer.hpp"

namespace {

constexpr std::uint64_t kItems = 50'000'000;
constexpr std::size_t kCapacity = 1024;
constexpr std::size_t kBatch = 64;
constexpr unsigned kProducerCore = 0;
constexpr unsigned kConsumerCore = 2;
constexpr int kRuns = 15;

using Clock = std::chrono::steady_clock;

double mops(Clock::duration elapsed) {
    return static_cast<double>(kItems) / std::chrono::duration<double>(elapsed).count() / 1e6;
}

double run_spsc_ring() {
    auto ch = spsc::channel<std::uint64_t>(kCapacity);
    auto rx = std::move(ch.second);
    std::barrier<> gate(2);
    std::thread producer([&gate, tx = std::move(ch.first)]() mutable {
        bench::pin_to_core(kProducerCore);
        gate.arrive_and_wait();
        for (std::uint64_t i = 0; i < kItems; ++i) {
            while (!tx.push(i)) {
            }
        }
    });
    bench::pin_to_core(kConsumerCore);
    gate.arrive_and_wait();
    const auto start = Clock::now();
    std::uint64_t received = 0;
    std::uint64_t value = 0;
    while (received < kItems) {
        if (rx.pop(value)) {
            ++received;
        }
    }
    const auto elapsed = Clock::now() - start;
    producer.join();
    return mops(elapsed);
}

double run_rigtorp() {
    rigtorp::SPSCQueue<std::uint64_t> queue(kCapacity);
    std::barrier<> gate(2);
    std::thread producer([&] {
        bench::pin_to_core(kProducerCore);
        gate.arrive_and_wait();
        for (std::uint64_t i = 0; i < kItems; ++i) {
            while (!queue.try_push(i)) {
            }
        }
    });
    bench::pin_to_core(kConsumerCore);
    gate.arrive_and_wait();
    const auto start = Clock::now();
    std::uint64_t received = 0;
    while (received < kItems) {
        if (queue.front() != nullptr) {
            queue.pop();
            ++received;
        }
    }
    const auto elapsed = Clock::now() - start;
    producer.join();
    return mops(elapsed);
}

double run_spsc_ring_batch() {
    auto ch = spsc::channel<std::uint64_t>(kCapacity);
    auto rx = std::move(ch.second);
    std::barrier<> gate(2);
    std::thread producer([&gate, tx = std::move(ch.first)]() mutable {
        bench::pin_to_core(kProducerCore);
        gate.arrive_and_wait();
        std::uint64_t chunk[kBatch];
        std::uint64_t next = 0;
        while (next < kItems) {
            const std::size_t want =
                static_cast<std::size_t>(std::min<std::uint64_t>(kBatch, kItems - next));
            for (std::size_t i = 0; i < want; ++i) {
                chunk[i] = next + i;
            }
            std::size_t sent = 0;
            while (sent < want) {
                const std::size_t n = tx.push_n(chunk + sent, want - sent);
                if (n == 0) {
                    _mm_pause();
                }
                sent += n;
            }
            next += want;
        }
    });
    bench::pin_to_core(kConsumerCore);
    gate.arrive_and_wait();
    const auto start = Clock::now();
    std::uint64_t received = 0;
    std::uint64_t buf[kBatch];
    while (received < kItems) {
        const std::size_t n = rx.pop_n(buf, kBatch);
        if (n == 0) {
            _mm_pause();
        }
        received += n;
    }
    const auto elapsed = Clock::now() - start;
    producer.join();
    return mops(elapsed);
}

void report(const char* name, double (*run)()) {
    std::vector<double> samples;
    samples.reserve(kRuns);
    for (int i = 0; i < kRuns; ++i) {
        samples.push_back(run());
    }
    std::printf("  %-22s %8.1f Melem/s\n", name, bench::median(std::move(samples)));
}

}  // namespace

int main() {
    std::printf("throughput (median of %d, %llu items, capacity %zu, cores %u->%u)\n", kRuns,
                static_cast<unsigned long long>(kItems), kCapacity, kProducerCore, kConsumerCore);
    report("spsc::Ring", run_spsc_ring);
    report("spsc::Ring batch(64)", run_spsc_ring_batch);
    report("rigtorp::SPSCQueue", run_rigtorp);
    std::printf("  (rigtorp::SPSCQueue has no batch API)\n");
    return 0;
}
